tests/phpunit/CRM/Profile/Form/EditTest.php

```diff
@@ -46,4 +46,67 @@ public function testProfileUrl(): void {
     $this->assertEquals('civicrm/Mr. Anthony Anderson II', CRM_Core_Session::singleton()->popUserContext());
   }
 
+  /**
+   * Test that requiring tags on a profile works.
+   *
+   * @throws \API_Exception
+   */
+  public function testProfileRequireTag(): void {
+    $ufGroupParams = [
+      'group_type' => 'Individual,Contact',
+      'name' => 'test_individual_contact_tag_profile',
+      'title' => 'Gimme a tag',
+      'api.uf_field.create' => [
+        [
+          'field_name' => 'first_name',
+          'is_required' => 1,
+          'visibility' => 'Public Pages and Listings',
+          'field_type' => 'Individual',
+          'label' => 'First Name',
+        ],
+        [
+          'field_name' => 'last_name',
+          'is_required' => 1,
+          'visibility' => 'Public Pages and Listings',
+          'field_type' => 'Individual',
+          'label' => 'Last Name',
+        ],
+        [
+          'field_name' => 'tag',
+          'is_required' => 1,
+          'visibility' => 'Public Pages and Listings',
+          'field_type' => 'Contact',
+          'label' => 'Tag',
+        ],
+      ]
+    ];
+
+    $profile = $this->callAPISuccess('uf_group', 'create', $ufGroupParams);
+    $profileID = $profile['id'];
+
+    // Configure the profile to be used as a standalone profile for data entry.
+    UFJoin::create(FALSE)->setValues([
+      'module' => 'Profile',
+      'uf_group_id' => $profileID,
+    ])->execute();
+
+    // Populate the form.
+    $formParams = [
+      'first_name' => 'Foo',
+      'last_name' => 'McGoo',
+      'gid' => $profileID,
+      'tag' => [],
+    ];
+    $form = $this->getFormObject('CRM_Profile_Form_Edit', $formParams);
+    $form->set('gid', $profileID);
+    $form->preProcess();
+    $form->buildQuickForm();
```

Comment:
I was thinking maybe adding $errors = $form->validate() here after buildQuickForm might do it but doesn't seem to. Hmm.

Reply:
I see it. This form handles its own error handling slightly differently than other forms (because it's Civi!) It expects there to be an errorUrl param and then does a redirect. The return value from validate() is just a true/false here, so it actually does work to call $form->validate(), I was just expecting an array of error messages. So you can do `$this->assertFalse($form->validate());` after the buildQuickForm(), and don't need the postProcess since then you're done with the test.